Potential regression: incomplete type in tag_of is a hard error

This issue was initially reported as https://github.com/boostorg/hana/issues/386. The following is a hard error in Boost trunk:

```c++
#include <boost/fusion/include/tag_of.hpp>
struct incomplete;
using X = ::boost::fusion::traits::tag_of<incomplete>::type;

int main() { }
```

I believe this started happening with https://github.com/boostorg/type_traits/commit/ac351390b215b73d8a9e16d2ca37ba2d9d88a14f. `fusion::tag_of` is using `boost::is_convertible`, which requires its input to be a complete type, but I don't think it is documented anywhere that tags must be complete types. Possible workarounds are
1. Only call `is_convertible` if `boost::is_complete` is true
2. Request that tags be complete types (but that's a breaking change)
3. Reimplement the `is_convertible` without the assertion

I think (1) is the best solution, even though I am skeptical that a `is_complete` trait can even be implemented without compiler support. But since Boost.TypeTraits provides it, it seems like it's possible.


The full error message (with Clang) is:

```
In file included from [snip].cpp:5:
In file included from [snip]/boost/fusion/include/tag_of.hpp:11:
In file included from [snip]/boost/fusion/support/tag_of.hpp:14:
In file included from [snip]/boost/fusion/support/detail/is_mpl_sequence.hpp:16:
[snip]/boost/type_traits/is_convertible.hpp:498:4: error: static_assert failed "From argument type to is_convertible must be a complete type"
   BOOST_STATIC_ASSERT_MSG(boost::is_complete<From>::value || boost::is_void<From>::value || boost::is_array<From>::value, "From argument type to is_convertible must be a complete type");
   ^                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[snip]/boost/static_assert.hpp:31:45: note: expanded from macro 'BOOST_STATIC_ASSERT_MSG'
#     define BOOST_STATIC_ASSERT_MSG( ... ) static_assert(__VA_ARGS__)
                                            ^             ~~~~~~~~~~~
[snip]/boost/mpl/aux_/nested_type_wknd.hpp:27:7: note: in instantiation of template class 'boost::is_convertible<incomplete, boost::fusion::detail::from_sequence_convertible_type>' requested here
    : T::type
      ^
[snip]/boost/mpl/not.hpp:41:11: note: in instantiation of template class 'boost::mpl::aux::nested_type_wknd<boost::is_convertible<incomplete, boost::fusion::detail::from_sequence_convertible_type> >' requested here
          BOOST_MPL_AUX_NESTED_TYPE_WKND(T)::value
          ^
[snip]/boost/mpl/aux_/nested_type_wknd.hpp:38:24: note: expanded from macro 'BOOST_MPL_AUX_NESTED_TYPE_WKND'
    ::boost::mpl::aux::nested_type_wknd<T> \
                       ^
[snip]/boost/mpl/aux_/nested_type_wknd.hpp:27:7: note: in instantiation of template class 'boost::mpl::not_<boost::is_convertible<incomplete, boost::fusion::detail::from_sequence_convertible_type> >' requested here
    : T::type
      ^
[snip]/boost/mpl/aux_/preprocessed/gcc/and.hpp:51:11: note: in instantiation of template class 'boost::mpl::aux::nested_type_wknd<boost::mpl::not_<boost::is_convertible<incomplete, boost::fusion::detail::from_sequence_convertible_type> > >' requested here
          BOOST_MPL_AUX_NESTED_TYPE_WKND(T1)::value
          ^
[snip]/boost/mpl/aux_/nested_type_wknd.hpp:38:24: note: expanded from macro 'BOOST_MPL_AUX_NESTED_TYPE_WKND'
    ::boost::mpl::aux::nested_type_wknd<T> \
                       ^
[snip]/boost/fusion/support/detail/is_mpl_sequence.hpp:22:11: note: in instantiation of template class 'boost::mpl::and_<boost::mpl::not_<boost::is_convertible<incomplete, boost::fusion::detail::from_sequence_convertible_type> >, boost::mpl::is_sequence<incomplete>, mpl_::bool_<true>, mpl_::bool_<true>, mpl_::bool_<true> >' requested here
        : mpl::and_<
          ^
[snip]/boost/mpl/if.hpp:63:68: note: in instantiation of template class 'boost::fusion::detail::is_mpl_sequence<incomplete>' requested here
          BOOST_MPL_AUX_STATIC_CAST(bool, BOOST_MPL_AUX_VALUE_WKND(T1)::value)
                                                                   ^
[snip]/boost/fusion/support/tag_of.hpp:53:18: note: in instantiation of template class 'boost::mpl::if_<boost::fusion::detail::is_mpl_sequence<incomplete>, boost::mpl::identity<boost::fusion::mpl_sequence_tag>, boost::mpl::identity<boost::fusion::non_fusion_tag> >' requested here
          : mpl::if_<fusion::detail::is_mpl_sequence<Sequence>,
                 ^
[snip]/boost/fusion/support/tag_of.hpp:71:15: note: in instantiation of template class 'boost::fusion::detail::tag_of_impl<incomplete, void>' requested here
            : boost::fusion::detail::tag_of_impl<Sequence, Active>
              ^
[snip].cpp:7:36: note: in instantiation of template class 'boost::fusion::traits::tag_of<incomplete, void>' requested here
using X = ::boost::fusion::traits::tag_of<incomplete>::type;
                                   ^
```

--- a/include/boost/fusion/support/detail/is_mpl_sequence.hpp
+++ b/include/boost/fusion/support/detail/is_mpl_sequence.hpp
@@ -13,6 +13,7 @@
 #include <boost/mpl/is_sequence.hpp>
 #include <boost/mpl/and.hpp>
 #include <boost/mpl/not.hpp>
+#include <boost/type_traits/is_complete.hpp>
 #include <boost/type_traits/is_convertible.hpp>
 
 namespace boost { namespace fusion { namespace detail
@@ -20,7 +21,7 @@ namespace boost { namespace fusion { namespace detail
     template <typename T>
     struct is_mpl_sequence
         : mpl::and_<
-            mpl::not_<is_convertible<T, from_sequence_convertible_type> >
+            mpl::not_<mpl::and_<is_complete<T>, is_convertible<T, from_sequence_convertible_type> > >
           , mpl::is_sequence<T> >
     {};
 }}}

--- a/include/boost/fusion/support/is_sequence.hpp
+++ b/include/boost/fusion/support/is_sequence.hpp
@@ -10,9 +10,11 @@
 #include <boost/fusion/support/config.hpp>
 #include <boost/fusion/support/sequence_base.hpp>
 #include <boost/fusion/support/tag_of.hpp>
+#include <boost/mpl/and.hpp>
+#include <boost/mpl/bool.hpp>
 #include <boost/mpl/is_sequence.hpp>
 #include <boost/mpl/or.hpp>
-#include <boost/mpl/bool.hpp>
+#include <boost/type_traits/is_complete.hpp>
 #include <boost/type_traits/is_convertible.hpp>
 #include <boost/type_traits/is_same.hpp>
 
@@ -69,7 +71,10 @@ namespace boost { namespace fusion
 
         template <typename Sequence, typename Enable = void>
         struct is_native_fusion_sequence
-          : is_convertible<Sequence, fusion::detail::from_sequence_convertible_type>
+          : mpl::and_<
+            is_complete<Sequence>,
+            is_convertible<Sequence, fusion::detail::from_sequence_convertible_type>
+          >
         {};
     }
 }}

--- a/test/Jamfile
+++ b/test/Jamfile
@@ -254,6 +254,7 @@ project
     [ run functional/invoke_procedure.cpp ]
     [ run sequence/swap.cpp ]
 
+    [ compile support/is_sequence.cpp ]
     [ compile support/pair_deque.cpp ]
     [ compile support/pair_list.cpp ]
     [ compile support/pair_map.cpp ]
@@ -264,6 +265,7 @@ project
         : [ requires cxx11_variadic_templates ] ]
     [ compile support/and.cpp
         : [ requires cxx11_variadic_templates ] ]
+    [ compile support/tag_of.cpp ]
 
 #   [ compile-fail xxx.cpp  ]
 

--- a/test/support/is_sequence.cpp
+++ b/test/support/is_sequence.cpp
@@ -1,0 +1,17 @@
+/*=============================================================================
+    Copyright (c) 2018 Louis Dionne
+
+    Distributed under the Boost Software License, Version 1.0. (See accompanying
+    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+==============================================================================*/
+#include <boost/fusion/support/is_sequence.hpp>
+#include <boost/static_assert.hpp>
+
+
+// Make sure fusion::is_sequence and fusion::is_native_fusion_sequence can be
+// used with an incomplete type.
+struct incomplete;
+BOOST_STATIC_ASSERT(!boost::fusion::traits::is_sequence<incomplete>::value);
+BOOST_STATIC_ASSERT(!boost::fusion::traits::is_native_fusion_sequence<incomplete>::value);
+
+int main() { }

--- a/test/support/tag_of.cpp
+++ b/test/support/tag_of.cpp
@@ -1,0 +1,17 @@
+/*=============================================================================
+    Copyright (c) 2018 Louis Dionne
+
+    Distributed under the Boost Software License, Version 1.0. (See accompanying
+    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+==============================================================================*/
+#include <boost/fusion/support/tag_of.hpp>
+#include <boost/static_assert.hpp>
+#include <boost/type_traits/is_same.hpp>
+
+
+// Make sure tag_of can be used with an incomplete type.
+struct incomplete;
+typedef boost::fusion::traits::tag_of<incomplete>::type Tag;
+BOOST_STATIC_ASSERT((boost::is_same<Tag, boost::fusion::non_fusion_tag>::value));
+
+int main() { }

